### PR TITLE
Add ButinaClustering estimator

### PR DESCRIPTION
**Pull request by rayair250-droid**
Part of #328 (Butina). Adds a scikit-learn-style `ButinaClustering` estimator, mirroring the existing `MaxMinClustering`.

### What it does
- `ButinaClustering(BaseEstimator, ClusterMixin)` with `_parameter_constraints`, `fit`, `predict`, `fit_predict`, `get_clusters_and_points`, and the same `centroid_indices_` / `centroid_bitvectors_` / `centroids_` / `labels_` attributes as `MaxMinClustering`.
- Clustering uses RDKit `Butina.ClusterData` on condensed lower-triangle Tanimoto distances (the RDKit-documented `isDistData=True` format), with `reordering=True` for deterministic output.
- Accepts the same inputs as `MaxMinClustering`: dense NumPy arrays, CSR sparse arrays, or sequences of `ExplicitBitVect`.
- `predict` assigns new samples to the nearest centroid by Tanimoto similarity.

### Notes / design choices to confirm
- I kept this to the new estimator only; the existing `butina_split` splitter is untouched (it can delegate to this class in a follow-up if you prefer, per the issue).
- `labels_` reflect Butina cluster membership (density/sphere-exclusion), which is not necessarily the nearest-centroid assignment — so unlike the MaxMin tests, I did not assert nearest-centroid for `fit` labels. `predict` (new samples) does use nearest-centroid.

### Verification
`ruff check` + `ruff format --check` clean; `pytest tests/clustering/` → **38 passed** (19 new for Butina, existing MaxMin unchanged). Verified against an editable install of the package.

Happy to adjust the API to match your intent for the scaffold/Butina classes.

### CODE REVIEW

File: skfp/clustering/butina.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+from sklearn.utils._param_validation import Interval, RealNotInt
+from sklearn.utils.validation import check_is_fitted, validate_data
+
+
+class ButinaClustering(BaseEstimator, ClusterMixin):
+    """
+    Taylor-Butina clustering.
+
+    A density-based clustering algorithm for binary fingerprints using Tanimoto
+    similarity, also known as sphere exclusion or leader-following clustering [1]_ [2]_.
```
Comment (j-adamczyk) [inline skfp/clustering/butina.py]: I would avoid this naming, the leader-following has certain differences, e.g. see slide 24 in Roger Sayle's talk

File: skfp/clustering/butina.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    Cluster centroids are chosen so that no two centroids are closer than a given
+    Tanimoto distance ``distance_threshold``; every remaining sample joins the cluster
+    of the first centroid within that radius. In contrast to centroid-based methods
+    like MaxMin clustering, Butina clusters follow the density of the data and can vary
+    widely in size.
+
+    Clustering uses RDKit ``Butina.ClusterData`` with ``reordering=True`` for
+    deterministic output. After fitting, new samples are assigned with :meth:`predict`
+    to the centroid with the highest Tanimoto similarity.
```
Comment (j-adamczyk) [inline skfp/clustering/butina.py]: Too much technical detail in docstring, remove

File: skfp/clustering/butina.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+    def _assign_labels(self, vectors: list[ExplicitBitVect]) -> np.ndarray:
+        """
+        Assign each sample to the nearest centroid by Tanimoto similarity.
+        """
+        labels = np.empty(len(vectors), dtype=int)
+        for i, fp in enumerate(vectors):
+            sims = BulkTanimotoSimilarity(fp, self.centroid_bitvectors_)
+            labels[i] = int(np.argmax(sims))
+        return labels
```
Comment (j-adamczyk) [inline skfp/clustering/butina.py]: 100% code duplication with MaxMin clustering. Create utils.py file in clustering and move those methods there as functions. Then also modify MaxMin clustering appropriately.

File: skfp/clustering/butina.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+    def get_clusters_and_points(self) -> dict[int, np.ndarray]:
+        """
+        Return a mapping from cluster id to the indices of its member samples.
+        """
+        check_is_fitted(self)
+        return {
+            k: np.where(self.labels_ == k)[0]
+            for k in range(len(self.centroid_indices_))
+        }
```
Comment (j-adamczyk) [inline skfp/clustering/butina.py]: If you rewrite this to use `self.labels_` directly, this can be moved to the parent class and be common for all clustering classes. This mapping does not require any notion of centroids, after all.

File: tests/clustering/butina.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    dense = np.asarray(binary_X.todense()) if hasattr(binary_X, "todense") else binary_X
+    labels_dense = ButinaClustering(distance_threshold=0.5).fit(dense).labels_
+    labels_sparse = (
+        ButinaClustering(distance_threshold=0.5).fit(csr_matrix(dense)).labels_
+    )
+    assert np.array_equal(labels_dense, labels_sparse)
+
+
+def test_predict_assigns_to_nearest_centroid(binary_X):
+    from rdkit.DataStructs import BulkTanimotoSimilarity
```
Comment (j-adamczyk) [inline tests/clustering/butina.py]: Import globally

**Review by j-adamczyk — CHANGES_REQUESTED** [review]

**Comment by rayair250-droid** [comment]
Thanks for the detailed review! Addressed everything in the latest push:

1. Dropped the `leader-following` naming.
2. Trimmed the extra technical paragraph from the docstring.
3. Moved `array_to_bitvectors` and `assign_labels` into `skfp/clustering/utils.py` and refactored **both** `ButinaClustering` and `MaxMinClustering` to use them (removed the duplicated private methods).
4. `get_clusters_and_points` now derives from `labels_` directly via a shared `utils.clusters_and_points` helper. I kept it as a utils function called by both classes rather than introducing a new base class, to stay minimal — happy to move it to a shared base class if you'd prefer that.
5. Global imports in the tests.

All 38 clustering tests pass; `ruff check`/`ruff format` clean.

File: skfp/clustering/butina.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+import numpy as np
+from rdkit.DataStructs import BulkTanimotoSimilarity
+from rdkit.DataStructs.cDataStructs import ExplicitBitVect
+from rdkit.ML.Cluster import Butina
+from scipy import sparse
+from sklearn.base import BaseEstimator, ClusterMixin
+from sklearn.utils._param_validation import Interval, RealNotInt
+from sklearn.utils.validation import check_is_fitted, validate_data
+
+from skfp.clustering import utils
```
Comment (j-adamczyk) [inline skfp/clustering/butina.py:12]: Import directly from utils, this is the convention that we use

File: skfp/clustering/utils.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
+import numpy as np
+from rdkit.DataStructs import BulkTanimotoSimilarity
+from rdkit.DataStructs.cDataStructs import ExplicitBitVect
+from scipy import sparse
+
+
+def array_to_bitvectors(X) -> list[ExplicitBitVect]:
+    """Convert input data to a list of RDKit ExplicitBitVect objects."""
```
Comment (j-adamczyk) [inline skfp/clustering/utils.py:8]: All docstrings use 3 separate lines, even oneliners, by convention

**Review by j-adamczyk — CHANGES_REQUESTED** [review]
